### ISSUE TITLE
NEC SOL

### ISSUE THREAD

**Issue by mergian** [pytorch-fdn/ecosystem#28]
### Contact emails

nicolas.weber@neclab.eu

### Project summary

SOL is an AI compiler specifically designed to be device and framework agnostic

### Project description

SOL is an AI compiler specifically designed to be device and framework agnostic. So far we support to execute PyTorch, TensorFlow, Keras and ONNX models, on X86 CPUs, NVIDIA GPUs and the NEC SX-Aurora. SOL supports to run inference, training and deploy models to standalone libraries. SOL integrates seamlessly into PyTorch and require only 2 lines of code: `import sol` and `model = sol.optimize(model)` or `model = torch.compile(model, backend='sol')`.

Some of the advanced features of SOL are:
- cross framework execution: enables to load TensorFlow, Keras or ONNX models and execute them in PyTorch, including training.
- foreign device support: SOL enables to run foreign hardware without native support within PyTorch ( https://arxiv.org/abs/2003.10688 )
- PyTorch identical pseudo random number generation, enabling SOL to be a 1:1 dropin replacement for PyTorch in training.

### Are there any other projects in the PyTorch Ecosystem similar to yours?  If, yes, what are they?

- PyTorch Inductor
- XLA
- GraphCore PopTorch
- NVIDIA TensorRT

### Project repo URL

https://github.com/SX-Aurora/veda-pytorch

### Additional repos in scope of the application

https://github.com/SX-Aurora/veda
https://github.com/SX-Aurora/veda-tensors

### Project license

3-BSD and propietary

### GitHub handles of the project maintainer(s)

mergian

### Is there a corporate or academic entity backing this project?  If so, please provide the name and URL of the entity.

NEC Laboratories Europe - https://www.neclab.eu/

### Website URL

https://sol.neclab.eu/

### Documentation

Documentation: https://sol.neclab.eu/docs/
Tutorials: https://sol.neclab.eu/docs/v0.7/tutorials.html

### How do you build and test the project today (continuous integration)?  Please describe.

We use an automated test-suite that runs models from torchvision, timm, transformers, ... and compares SOL outputs and gradients for numerical accuracy against PyTorch. Additional tests are available to brute force test all parameter combinations for layers.

Our recent SOL v0.7.1 release supports 84% of torchvision, 92% of timm and 52% of transformers.

### Version of PyTorch

Current SOL supports all PyTorch since v1.10.1.

### Components of PyTorch

- ATen and C10 device support # to use PyTorch's memory allocation system, e.g., to prevent out-of-memory on NVIDIA as PyTorch might preallocate a huge chunk of the GPU's memory.
- torch.compile # for integration into torch.compile(..., backend='sol')
- torch.fx # for parsing torch.fx.GraphModule models
- torch.jit # for legacy scripted models

### How long do you expect to maintain the project?

No project end defined or to be expected any time soon.

### Additional information

SOL is split into multiple subprojects:
- Our PyTorch extensions that add the required NEC SX-Aurora functionality to PyTorch are open source (3-BSD license), see linked repositories.
- The core SOL compiler is a propietary software (similar to NVIDIA's TensorRT or GraphCore's compiler). It's free to use for users of the NEC SX-Aurora. Licenses to other device backends can be acquired, see https://neclab.eu/services/sol/sol-for-ai-developers )

We understand that the PyTorch landscape mainly aims for pure open source projects, but as the core compilers of PopTorch and TensorRT are also closed source, we think SOL as compiler that enables to run PyTorch on the NEC SX-Aurora, it should also be mentioned. Similar to PopTorch and TensorRT, the public APIs for the NEC SX-Aurora are open source, see the linked repositories.

If required we can do a demo for the PyTorch team to showcase our functionality.

To clarify the naming of "VEDA" and "SOL". VEDA are the AI extensions to integrate the NEC SX-Aurora (also called Vector Engine "VE") into PyTorch. But as SOL is multi-device and is only not limited to the NEC SX-Aurora, it has a different name. And no, SOL is no acronym, it's the latin word for "sun" ;)

**Comment by jspisak** [comment]
is that the correct repo provided above? It doesn't seem very active with the last commit 7 months ago and only 4 stars on the repo. Do we have the correct repo?

**Comment by mergian** [comment]
@lantiga thanks, I understand. Sure, how can we get in touch with the working group?
